File: src/main/java/com/lm_pakkanen/radio_pele_java/controllers/commands/ShuffleCommand.java
package com.lm_pakkanen.radio_pele_java.controllers.commands;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Component;

import com.lm_pakkanen.radio_pele_java.controllers.MailMan;
import com.lm_pakkanen.radio_pele_java.controllers.TrackScheduler;
import com.lm_pakkanen.radio_pele_java.interfaces.ICommandListener;
import com.lm_pakkanen.radio_pele_java.models.exceptions.InvalidChannelException;

import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData;

@Component
public final class ShuffleCommand extends BaseCommand
    implements ICommandListener {

  private @NonNull TrackScheduler trackScheduler;

  public ShuffleCommand(@Autowired @NonNull TrackScheduler trackScheduler) {
    super();
    this.trackScheduler = trackScheduler;
  }

  @Override
  public String getCommandName() {
    return "shuffle";
  }

  @Override
  public String getCommandDescription() {
    return "Shuffle the Q.";
  }

  @Override
  public SlashCommandData getCommandData() {
    return Commands.slash(this.getCommandName(), this.getCommandDescription());
  }

  /**
   * Shuffles the current queue.
   */
  @Override
  public void onSlashCommandInteraction(SlashCommandInteractionEvent event)
      throws NullPointerException {
    if (!event.getName().equals(this.getCommandName())) {
      return;
    }

    try {
      super.getTextChan(event);
      this.trackScheduler.shuffle();
      MailMan.replyInteractionMessage(event, "Q shuffled.");
    } catch (InvalidChannelException exception) {
      String exceptionMessage = exception.getMessage();

      if (exceptionMessage == null) {
        exceptionMessage = "Unknown exception occurred.";
      }

      MailMan.replyInteractionMessage(event, exceptionMessage);
    }
  }
}

File: bot/src/main/kotlin/com/lmpakkanen/radiopelejava/commands/ShuffleCommand.kt
package com.lmpakkanen.radiopelejava.commands

import com.lmpakkanen.radiopelejava.controllers.MailMan
import com.lmpakkanen.radiopelejava.controllers.TrackScheduler
import com.lmpakkanen.radiopelejava.interfaces.ICommandListener
import com.lmpakkanen.radiopelejava.models.exceptions.InvalidChannelException
import com.lmpakkanen.radiopelejava.models.messages.embeds.ExceptionEmbed
import com.lmpakkanen.radiopelejava.models.messages.embeds.QueueShuffledEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.build.Commands
import org.springframework.stereotype.Component

@Component
class ShuffleCommand(
    private val trackScheduler: TrackScheduler,
) : BaseCommand(),
    ICommandListener {
    override val commandName = "shuffle"
    override val commandDescription = "Shuffle the Q."
    override val commandData = Commands.slash(this.commandName, this.commandDescription)

    /**
     * Shuffles the current queue.
     *
     * @throws NullPointerException
     */
    @Throws(NullPointerException::class)
    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
        if (event.name != this.commandName) {
            return
        }

        try {
            super.getTextChan(event)
            this.trackScheduler.shuffle()
            MailMan.replyInteraction(event, QueueShuffledEmbed().embed)
        } catch (exception: InvalidChannelException) {
            MailMan.replyInteraction(event, ExceptionEmbed(exception).embed)
        }
    }
}
